### PR TITLE
Calculator Project (Half Platinum)

### PR DESCRIPTION
**Pull request by catchingash**
This is my gold/dinner + half platinum/dessert calculator. The program returns floats when the number needs to be a float (e.g. 4.5), and returns integers when it's a whole number (e.g. 4.0 --> 4). However, it is missing the ability to do multiple operations / handle parenthesis.

### CODE REVIEW

**Comment by dougcole** [comment]
Reviewing this now.

File: exercise_150507_Calculator.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+subtract = ["substraction", "sub", "subtract", "-"]
+add = ["addition", "add", "+"]
+multiply = ["multiplication", "multiply", "x", "*"]
+divide = ["division", "div", "/"]
+exponent = ["exponential", "exponents", "exponent", "exp", "power", "^", "**"]
+sqrt = ["squareroot", "square root", "sqrt", "^1/2", "\u221A"]
+mod = ["remainders", "remainder", "modulo", "modulus", "%"]
+operations = subtract + add + multiply + divide + exponent + sqrt + mod
+UNACCEPTED_OPERATION_MEMO = "I'm sorry. I'm confused. I know it's probably obvious to a human, but could you try again and dumb it down for me? (You'll need to restart me. Sorry!)"
+UNEXPECTED_ERROR_MEMO = "I'm not sure what happened. I unencountered an unexpected error and had to close. I'm sorry!"
```
Comment (dougcole) [inline exercise_150507_Calculator.rb:14]: I like the extraction of constants into named variables at the top - it makes the rest of the code read much better.

File: exercise_150507_Calculator.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+while operator == "help"
+  puts "Thanks for asking. I do lots of math! Your options are:\naddition, subtraction, multiplication, division, exponents, square root, remainder."
+  operator = gets.chomp.downcase.strip
+end
+
+abort(UNACCEPTED_OPERATION_MEMO) unless operations.include?(operator)
+
+# Get user input: number selection
+puts "Exciting! Pick the first number."
+num1 = gets.chomp.delete(",")
```
Comment (dougcole) [inline exercise_150507_Calculator.rb:30]: I'm curious - what prompted the `delete(",")`?

File: exercise_150507_Calculator.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+# Get user input: 2nd number (if needed)
+unless sqrt.include?(operator)
+  puts "I can work with that. What's the second number?"
+  num2 = gets.chomp.delete(",")
+  if num2 == "0"
+  elsif num2.to_f == 0
+    puts "You're funny. That's not a number! I'll just pretend you picked 0."
+  end
+  num2 = num2.to_f
```
Comment (dougcole) [inline exercise_150507_Calculator.rb:45]: The `num2 = gets.chomp.delete(",")` to this line is the same as line 30-35, you could remove that duplication by defining a function and re-using it in both locations.

File: exercise_150507_Calculator.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  num_calculated = num1 ** num2
+  operator = "^"
+elsif sqrt.include?(operator) ## sqrt will always be reported as a float.
+  num_calculated = Math.sqrt(num1.abs)
+  operator = "\u221A"
+  abort("#{operator}#{num1} = #{num_calculated}\nI'm so awesome! Run me again to do more math!")
+elsif mod.include?(operator)
+  num_calculated = num1 % num2
+  operator = "%"
+end
```
Comment (dougcole) [inline exercise_150507_Calculator.rb:71]: I like how you separated out the calculation of the result here from the display below, makes it easy to follow and modify!

File: exercise_150507_Calculator.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+num1 = num1.to_i if (num1 % 1) == 0
+num2 = num2.to_i if (num2 % 1) == 0
+
+if (num_calculated % 1) == 0
+  puts "#{num1} #{operator} #{num2} = #{num_calculated.to_i}\nI'm so awesome! Run me again to do more math!" # TODO: re-prompt instead of ending?
+elsif (num_calculated % 1) != 0
+  puts "#{num1} #{operator} #{num2} = #{num_calculated}\nI'm so awesome! Run me again to do more math!" # TODO: Add functionality to round? User-selected rounding?
+else
+  abort(UNEXPECTED_ERROR_MEMO)
+end
```
Comment (dougcole) [inline exercise_150507_Calculator.rb:83]: I'd lean towards moving the puts out of the conditional, as it looks like the only difference is that in the case of `(num_calculated % 1) == 0` you set `num_calculated = num_calculated.to_i`.

File: exercise_150507_Calculator.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+# Remove *.0 from the numbers (if exists); Provide output to the user
+num1 = num1.to_i if (num1 % 1) == 0
+num2 = num2.to_i if (num2 % 1) == 0
+
+if (num_calculated % 1) == 0
+  puts "#{num1} #{operator} #{num2} = #{num_calculated.to_i}\nI'm so awesome! Run me again to do more math!" # TODO: re-prompt instead of ending?
+elsif (num_calculated % 1) != 0
+  puts "#{num1} #{operator} #{num2} = #{num_calculated}\nI'm so awesome! Run me again to do more math!" # TODO: Add functionality to round? User-selected rounding?
+else
+  abort(UNEXPECTED_ERROR_MEMO)
```
Comment (dougcole) [inline exercise_150507_Calculator.rb:82]: This seems worth a comment as it's not immediately clear what case this protects against.

**Comment by dougcole** [comment]
This is great! Easy to read and well documented, let me know if you have any questions on my comments or any other questions about the project!

File: exercise_150507_Calculator.rb (hunk 3 above)
Comment (dougcole) [inline exercise_150507_Calculator.rb:45]: Just realized you may not have covered writing functions yet! Feel free to ignore this comment. :smile:

File: exercise_150507_Calculator.rb (hunk 2 above)
Comment (catchingash) [inline exercise_150507_Calculator.rb:30]: I sent my calculator to a friend, who promptly put in "5,000" as one of the numbers. "5,000" was treated as though it was "5" instead. I added .delete(",") so it would be treated the same as "5000".

File: exercise_150507_Calculator.rb (hunk 3 above)
Comment (catchingash) [inline exercise_150507_Calculator.rb:45]: We hadn't when I wrote that code, but we have now. Defining a function would've been a much cleaner way to do it!

File: exercise_150507_Calculator.rb (hunk 4 above)
Comment (catchingash) [inline exercise_150507_Calculator.rb:71]: Thanks :)

File: exercise_150507_Calculator.rb (hunk 5 above)
Comment (catchingash) [inline exercise_150507_Calculator.rb:83]: Your suggestion definitely makes more sense, given where I ended. I was afraid I was going to need my num_calculated as a float for later functionality, but never got around to it. (In retrospect, I could've converted it back, anyway.)

File: exercise_150507_Calculator.rb (hunk 6 above)
Comment (catchingash) [inline exercise_150507_Calculator.rb:82]: Just me being nervous that somehow my if/elsif didn't catch all possible conditions. Completely un-necessary, in retrospect.

**Comment by catchingash** [comment]
Thanks! I really appreciate all of your feedback :+1:, and it was all clear/easy-to-understand. No questions from me :)

File: exercise_150507_Calculator.rb (hunk 2 above)
Comment (dougcole) [inline exercise_150507_Calculator.rb:30]: oh! That makes total sense, I hadn't even thought of it. Nice!
